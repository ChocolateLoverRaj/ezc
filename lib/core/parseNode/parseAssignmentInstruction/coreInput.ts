import coreParseCallAssignableInput from '../parseCallAssignable/coreInput'
import parseCallAssignable from '../parseCallAssignable/parseCallAssignable'
import Input from './Input'

const coreInput: Input = [parseCallAssignable(coreParseCallAssignableInput)]

export default coreInput
