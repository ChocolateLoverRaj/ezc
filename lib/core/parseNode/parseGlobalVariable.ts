import CoreTokenDatas from '../CoreTokenDatas'
import CoreTokenType from '../CoreTokenType'
import KeyWord from '../KeyWord'
import skip from '../splitAsyncIterator/skip'
import splitAsyncIterator from '../splitAsyncIterator/splitAsyncIterator'
import ConstantOrGlobal from './ConstantOrGlobal'
import CoreNodesWithData from './CoreNodesWithData'
import CoreNodeType from './CoreNodeType'
import coreTypeParsers from './coreTypeParsers'
import coreValueParsers from './coreValueParsers'
import Linkage from './Linkage'
import parseIdentifier from './parseIdentifier'
import tryNodeParsers from './tryNodeParsers'
import TryParseNode from './TryParseNode'

const parseGlobalVariable: TryParseNode<CoreNodesWithData[CoreNodeType.GLOBAL_VARIABLE]> =
async stream => {
  const asyncIterator = stream[Symbol.asyncIterator]()
  const splittedAsyncIterator = splitAsyncIterator(asyncIterator)
  const identifier = await parseIdentifier(splittedAsyncIterator.asyncIterable)
  if (identifier === undefined) return
  skip(splittedAsyncIterator, identifier.length)

  {
    const { done, value } = await splittedAsyncIterator.asyncIterable[Symbol.asyncIterator]().next()
    if (done === true) return
    if (!(value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD)) return
    const data = value.data as CoreTokenDatas[CoreTokenType.KEY_WORD]
    if (!(data.enum === KeyWord && data.id === KeyWord.EQUALS)) return
  }
  skip(splittedAsyncIterator, 1)

  const allowedKeyWords = [
    KeyWord.PRIVATE,
    KeyWord.CONSTANT,
    KeyWord.GLOBAL,
    KeyWord.UNNAMED_ADDR
  ]
  const keyWords: KeyWord[] = []
  for await (const value of splittedAsyncIterator.asyncIterable) {
    if (value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD) {
      const data = value.data as CoreTokenDatas[CoreTokenType.KEY_WORD]
      if (!(data.enum === KeyWord && allowedKeyWords.includes(data.id))) return
      keyWords.push(data.id)
    } else {
      break
    }
  }
  skip(splittedAsyncIterator, keyWords.length)
  const isConstant = keyWords.includes(KeyWord.CONSTANT)
  const isGlobal = keyWords.includes(KeyWord.GLOBAL)
  // Must have exactly one of constant or global
  if (Number(isConstant) + Number(isGlobal) !== 1) {
    return
  }

  const type = await tryNodeParsers(
    coreTypeParsers)(splittedAsyncIterator.asyncIterable)
  if (type === undefined) return
  skip(splittedAsyncIterator, type.length)

  const value = await tryNodeParsers(coreValueParsers)(splittedAsyncIterator.asyncIterable)
  if (value === undefined) return
  skip(splittedAsyncIterator, value.length)

  return {
    node: {
      type: {
        enum: CoreNodeType,
        id: CoreNodeType.GLOBAL_VARIABLE
      },
      data: {
        // FIXME: No align logic
        align: undefined,
        constantOrGlobal: isConstant
          ? ConstantOrGlobal.CONSTANT
          : ConstantOrGlobal.GLOBAL,
        identifier: identifier.node,
        // FIXME: No other linkage types
        linkage: Linkage.PRIVATE,
        type: type.node,
        unnamed_addr: keyWords.includes(KeyWord.UNNAMED_ADDR),
        value: value.node
      }
    },
    length: identifier.length + 1 + keyWords.length + type.length + value.length
  }
}

export default parseGlobalVariable
