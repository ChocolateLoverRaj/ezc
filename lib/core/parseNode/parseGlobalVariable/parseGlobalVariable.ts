import CoreTokenDatas from '../../CoreTokenDatas'
import CoreTokenType from '../../CoreTokenType'
import CoreKeyWord from '../../CoreKeyWord'
import skip from '../../splitAsyncIterator/skip'
import splitAsyncIterator from '../../splitAsyncIterator/splitAsyncIterator'
import ConstantOrGlobal from '../ConstantOrGlobal'
import CoreNodesWithData from '../CoreNodesWithData'
import CoreNodeType from '../CoreNodeType'
import Linkage from '../Linkage'
import parseIdentifier from '../parseIdentifier'
import tryNodeParsers from '../tryNodeParsers'
import TryParseNode from '../TryParseNode'
import Input from './Input'

const parseGlobalVariable = (
  { typeParsers, valueParsers }: Input
): TryParseNode<CoreNodesWithData[CoreNodeType.GLOBAL_VARIABLE]> =>
  async stream => {
    const asyncIterator = stream[Symbol.asyncIterator]()
    const splittedAsyncIterator = splitAsyncIterator(asyncIterator)
    const identifier = await parseIdentifier(splittedAsyncIterator.asyncIterable)
    if (identifier === undefined) return
    skip(splittedAsyncIterator, identifier.length)

    {
      const { done, value } =
        await splittedAsyncIterator.asyncIterable[Symbol.asyncIterator]().next()
      if (done === true) return
      if (!(value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD)) return
      const data = value.data as CoreTokenDatas[CoreTokenType.KEY_WORD]
      if (!(data.enum === CoreKeyWord && data.id === CoreKeyWord.EQUALS)) return
    }
    skip(splittedAsyncIterator, 1)

    const allowedKeyWords = [
      CoreKeyWord.PRIVATE,
      CoreKeyWord.CONSTANT,
      CoreKeyWord.GLOBAL,
      CoreKeyWord.UNNAMED_ADDR
    ]
    const keyWords: CoreKeyWord[] = []
    for await (const value of splittedAsyncIterator.asyncIterable) {
      if (value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD) {
        const data = value.data as CoreTokenDatas[CoreTokenType.KEY_WORD]
        if (!(data.enum === CoreKeyWord && allowedKeyWords.includes(data.id))) return
        keyWords.push(data.id)
      } else {
        break
      }
    }
    skip(splittedAsyncIterator, keyWords.length)
    const isConstant = keyWords.includes(CoreKeyWord.CONSTANT)
    const isGlobal = keyWords.includes(CoreKeyWord.GLOBAL)
    // Must have exactly one of constant or global
    if (Number(isConstant) + Number(isGlobal) !== 1) {
      return
    }

    const type = await tryNodeParsers(typeParsers)(splittedAsyncIterator.asyncIterable)
    if (type === undefined) return
    skip(splittedAsyncIterator, type.length)

    const value = await tryNodeParsers(valueParsers)(splittedAsyncIterator.asyncIterable)
    if (value === undefined) return
    skip(splittedAsyncIterator, value.length)

    return {
      node: {
        type: {
          enum: CoreNodeType,
          id: CoreNodeType.GLOBAL_VARIABLE
        },
        data: {
        // FIXME: No align logic
          align: undefined,
          constantOrGlobal: isConstant
            ? ConstantOrGlobal.CONSTANT
            : ConstantOrGlobal.GLOBAL,
          identifier: identifier.node,
          // FIXME: No other linkage types
          linkage: Linkage.PRIVATE,
          type: type.node,
          unnamed_addr: keyWords.includes(CoreKeyWord.UNNAMED_ADDR),
          value: value.node
        }
      },
      length: identifier.length + 1 + keyWords.length + type.length + value.length
    }
  }

export default parseGlobalVariable
