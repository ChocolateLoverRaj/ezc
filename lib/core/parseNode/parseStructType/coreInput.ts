import coreTypeParsers from '../coreTypeParsers'
import Input from './Input'

const coreInput: Input = coreTypeParsers

export default coreInput
