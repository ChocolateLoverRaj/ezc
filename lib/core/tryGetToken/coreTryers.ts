import CoreTokenWithData from '../CoreTokenWithData'
import coreParseKeywordOptions from './coreParseKeywordOptions'
import parseIdentifier from './parseIdentifier'
import parseIntegerType from './parseIntegerType'
import parseKeyword from './parseKeyword'
import parseNumberLiteral from './parseNumberLiteral'
import parseOpenClose from './parseOpenClose'
import parseStringLiteral from './parseStringLiteral'
import TryGetToken from './TryGetToken'

const coreTryers: ReadonlyArray<TryGetToken<CoreTokenWithData>> = [
  parseIdentifier,
  parseIntegerType,
  parseKeyword(coreParseKeywordOptions),
  parseNumberLiteral,
  parseOpenClose,
  parseStringLiteral
] as const

export default coreTryers
