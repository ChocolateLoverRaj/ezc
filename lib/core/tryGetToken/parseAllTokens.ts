import EnumItemWithData from '../EnumItemWithData'
import skip from '../splitAsyncIterator/skip'
import splitAsyncIterator from '../splitAsyncIterator/splitAsyncIterator'
import parseToken from './parseToken'
import TryGetToken from './TryGetToken'

const parseAllTokens = (tryers: ReadonlyArray<TryGetToken<EnumItemWithData>>) => async function * (
  asyncIterable: AsyncIterable<string>
): AsyncGenerator<EnumItemWithData | undefined> {
  const splittedIterator = splitAsyncIterator(asyncIterable[Symbol.asyncIterator]())
  let chunkBefore = ''
  while (true) {
    const token = await parseToken(tryers)({
      async * [Symbol.asyncIterator] () {
        yield chunkBefore
        for await (const chunk of splittedIterator.asyncIterable) {
          yield chunk
        }
      }
    })
    if (token === undefined) {
      yield undefined
      return
    }
    yield token.token
    let chunksProcessed = 0
    let charactersProcessed = 0
    if (token.length < chunkBefore.length) {
      chunkBefore = chunkBefore.slice(token.length - chunkBefore.length)
    } else {
      for await (const chunk of splittedIterator.asyncIterable) {
        chunksProcessed++
        charactersProcessed += chunk.length
        if (charactersProcessed >= token.length) {
          chunkBefore = charactersProcessed === token.length
            ? ''
            : chunk.slice(token.length - charactersProcessed)
          break
        }
      }
      if (chunksProcessed === 0) return
      skip(splittedIterator, chunksProcessed)
    }
  }
}

export default parseAllTokens
