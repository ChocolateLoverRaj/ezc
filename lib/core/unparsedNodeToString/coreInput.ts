import coreUnparseTokenInput from '../unparseToken/unparseToken/coreInput'
import coreNodeUnparsers from './coreNodeUnparsers'
import Input from './Input'

const coreInput: Input = {
  unparseTokenInput: coreUnparseTokenInput,
  nodeUnparsers: coreNodeUnparsers
}

export default coreInput
