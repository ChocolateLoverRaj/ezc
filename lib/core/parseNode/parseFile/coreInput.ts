import coreParseDeclareInput from '../parseDeclare/coreInput'
import parseDeclare from '../parseDeclare/parseDeclare'
import parseFunction from '../parseFunction/parseFunction'
import coreParseGlobalVariableInput from '../parseGlobalVariable/coreInput'
import parseGlobalVariable from '../parseGlobalVariable/parseGlobalVariable'
import tryNodeParsers from '../tryNodeParsers'
import Input from './Input'
import coreParseFunctionInput from '../parseFunction/coreInput'

const coreInput: Input = tryNodeParsers([
  parseDeclare(coreParseDeclareInput),
  parseGlobalVariable(coreParseGlobalVariableInput),
  parseFunction(coreParseFunctionInput)
])

export default coreInput
