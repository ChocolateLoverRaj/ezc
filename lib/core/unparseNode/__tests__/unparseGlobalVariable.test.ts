import arrayToAsyncIterable from '../../arrayToAsyncIterable'
import coreInput from '../../parseNode/parseGlobalVariable/coreInput'
import parseGlobalVariable from '../../parseNode/parseGlobalVariable/parseGlobalVariable'
import coreTryers from '../../tryGetToken/coreTryers'
import parseAllTokens from '../../tryGetToken/parseAllTokens'
import unparseGlobalVariable from '../unparseGlobalVariable'

test('@0 = private unnamed_addr constant [13 x i8] c"Hello World!\\00"', async () => {
  expect(unparseGlobalVariable((await parseGlobalVariable(coreInput)(parseAllTokens(coreTryers)(
    arrayToAsyncIterable([
      '@0 = private unnamed_addr constant [13 x i8] c"Hello World!\\00"'
    ])
  ) as any))?.node.data as any)).toMatchSnapshot()
})
