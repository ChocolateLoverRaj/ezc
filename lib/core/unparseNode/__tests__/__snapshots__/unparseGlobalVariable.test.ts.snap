// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@0 = private unnamed_addr constant [13 x i8] c"Hello World!\\00" 1`] = `
Array [
  Object {
    "data": Object {
      "data": Object {
        "name": "0",
        "type": 0,
      },
      "type": Object {
        "enum": Object {
          "0": "FILE",
          "1": "DECLARE",
          "10": "IDENTIFIER",
          "11": "NUMBER",
          "12": "VOID_TYPE",
          "13": "BLOCK",
          "14": "RETURN_INSTRUCTION",
          "15": "CALL_ASSIGNABLE",
          "16": "ASSIGNMENT_INSTRUCTION",
          "17": "INPUT",
          "18": "INPUT_FLAG",
          "2": "FUNCTION",
          "3": "INTEGER_TYPE",
          "4": "GLOBAL_VARIABLE",
          "5": "STRING",
          "6": "FLOAT_TYPE",
          "7": "POINTER_TYPE",
          "8": "ARRAY_TYPE",
          "9": "STRUCT_TYPE",
          "ARRAY_TYPE": 8,
          "ASSIGNMENT_INSTRUCTION": 16,
          "BLOCK": 13,
          "CALL_ASSIGNABLE": 15,
          "DECLARE": 1,
          "FILE": 0,
          "FLOAT_TYPE": 6,
          "FUNCTION": 2,
          "GLOBAL_VARIABLE": 4,
          "IDENTIFIER": 10,
          "INPUT": 17,
          "INPUT_FLAG": 18,
          "INTEGER_TYPE": 3,
          "NUMBER": 11,
          "POINTER_TYPE": 7,
          "RETURN_INSTRUCTION": 14,
          "STRING": 5,
          "STRUCT_TYPE": 9,
          "VOID_TYPE": 12,
        },
        "id": 10,
      },
    },
    "type": 2,
  },
  Object {
    "data": undefined,
    "type": 1,
  },
  Object {
    "data": Object {
      "data": Object {
        "enum": Object {
          "0": "PRIVATE",
          "1": "UNNAMED_ADDR",
          "10": "RETURN",
          "11": "EQUALS",
          "12": "GLOBAL",
          "13": "NO_CAPTURE",
          "14": "NO_ALIAS",
          "15": "COMMA",
          "16": "VOID",
          "17": "NO_FREE",
          "18": "CALL",
          "19": "OPEN_PARENTHESIS",
          "2": "CONSTANT",
          "20": "CLOSE_PARENTHESIS",
          "21": "OPEN_BRACKET",
          "22": "CLOSE_BRACKET",
          "23": "OPEN_CURLY_BRACKET",
          "24": "CLOSE_CURLY_BRACKET",
          "25": "HALF",
          "26": "BFLOAT",
          "27": "FLOAT",
          "28": "DUOBLE",
          "29": "FP128",
          "3": "ALIGN",
          "30": "X86_FP80",
          "31": "PPC_FP128",
          "4": "DECLARE",
          "5": "DEFINE",
          "6": "PTR",
          "7": "GET_ELEMENT_PTR",
          "8": "INBOUNDS",
          "9": "X",
          "ALIGN": 3,
          "BFLOAT": 26,
          "CALL": 18,
          "CLOSE_BRACKET": 22,
          "CLOSE_CURLY_BRACKET": 24,
          "CLOSE_PARENTHESIS": 20,
          "COMMA": 15,
          "CONSTANT": 2,
          "DECLARE": 4,
          "DEFINE": 5,
          "DUOBLE": 28,
          "EQUALS": 11,
          "FLOAT": 27,
          "FP128": 29,
          "GET_ELEMENT_PTR": 7,
          "GLOBAL": 12,
          "HALF": 25,
          "INBOUNDS": 8,
          "NO_ALIAS": 14,
          "NO_CAPTURE": 13,
          "NO_FREE": 17,
          "OPEN_BRACKET": 21,
          "OPEN_CURLY_BRACKET": 23,
          "OPEN_PARENTHESIS": 19,
          "PPC_FP128": 31,
          "PRIVATE": 0,
          "PTR": 6,
          "RETURN": 10,
          "UNNAMED_ADDR": 1,
          "VOID": 16,
          "X": 9,
          "X86_FP80": 30,
        },
        "id": 11,
      },
      "type": Object {
        "enum": Object {
          "0": "IDENTIFIER",
          "1": "NUMBER_LITERAL",
          "2": "INTEGER_TYPE",
          "3": "STRING_LITERAL",
          "4": "KEY_WORD",
          "IDENTIFIER": 0,
          "INTEGER_TYPE": 2,
          "KEY_WORD": 4,
          "NUMBER_LITERAL": 1,
          "STRING_LITERAL": 3,
        },
        "id": 4,
      },
    },
    "type": 0,
  },
  Object {
    "data": undefined,
    "type": 1,
  },
  Object {
    "data": Object {
      "data": Object {
        "enum": Object {
          "0": "PRIVATE",
          "1": "UNNAMED_ADDR",
          "10": "RETURN",
          "11": "EQUALS",
          "12": "GLOBAL",
          "13": "NO_CAPTURE",
          "14": "NO_ALIAS",
          "15": "COMMA",
          "16": "VOID",
          "17": "NO_FREE",
          "18": "CALL",
          "19": "OPEN_PARENTHESIS",
          "2": "CONSTANT",
          "20": "CLOSE_PARENTHESIS",
          "21": "OPEN_BRACKET",
          "22": "CLOSE_BRACKET",
          "23": "OPEN_CURLY_BRACKET",
          "24": "CLOSE_CURLY_BRACKET",
          "25": "HALF",
          "26": "BFLOAT",
          "27": "FLOAT",
          "28": "DUOBLE",
          "29": "FP128",
          "3": "ALIGN",
          "30": "X86_FP80",
          "31": "PPC_FP128",
          "4": "DECLARE",
          "5": "DEFINE",
          "6": "PTR",
          "7": "GET_ELEMENT_PTR",
          "8": "INBOUNDS",
          "9": "X",
          "ALIGN": 3,
          "BFLOAT": 26,
          "CALL": 18,
          "CLOSE_BRACKET": 22,
          "CLOSE_CURLY_BRACKET": 24,
          "CLOSE_PARENTHESIS": 20,
          "COMMA": 15,
          "CONSTANT": 2,
          "DECLARE": 4,
          "DEFINE": 5,
          "DUOBLE": 28,
          "EQUALS": 11,
          "FLOAT": 27,
          "FP128": 29,
          "GET_ELEMENT_PTR": 7,
          "GLOBAL": 12,
          "HALF": 25,
          "INBOUNDS": 8,
          "NO_ALIAS": 14,
          "NO_CAPTURE": 13,
          "NO_FREE": 17,
          "OPEN_BRACKET": 21,
          "OPEN_CURLY_BRACKET": 23,
          "OPEN_PARENTHESIS": 19,
          "PPC_FP128": 31,
          "PRIVATE": 0,
          "PTR": 6,
          "RETURN": 10,
          "UNNAMED_ADDR": 1,
          "VOID": 16,
          "X": 9,
          "X86_FP80": 30,
        },
        "id": 0,
      },
      "type": Object {
        "enum": Object {
          "0": "IDENTIFIER",
          "1": "NUMBER_LITERAL",
          "2": "INTEGER_TYPE",
          "3": "STRING_LITERAL",
          "4": "KEY_WORD",
          "IDENTIFIER": 0,
          "INTEGER_TYPE": 2,
          "KEY_WORD": 4,
          "NUMBER_LITERAL": 1,
          "STRING_LITERAL": 3,
        },
        "id": 4,
      },
    },
    "type": 0,
  },
  Object {
    "data": undefined,
    "type": 1,
  },
  Object {
    "data": Object {
      "data": Object {
        "enum": Object {
          "0": "PRIVATE",
          "1": "UNNAMED_ADDR",
          "10": "RETURN",
          "11": "EQUALS",
          "12": "GLOBAL",
          "13": "NO_CAPTURE",
          "14": "NO_ALIAS",
          "15": "COMMA",
          "16": "VOID",
          "17": "NO_FREE",
          "18": "CALL",
          "19": "OPEN_PARENTHESIS",
          "2": "CONSTANT",
          "20": "CLOSE_PARENTHESIS",
          "21": "OPEN_BRACKET",
          "22": "CLOSE_BRACKET",
          "23": "OPEN_CURLY_BRACKET",
          "24": "CLOSE_CURLY_BRACKET",
          "25": "HALF",
          "26": "BFLOAT",
          "27": "FLOAT",
          "28": "DUOBLE",
          "29": "FP128",
          "3": "ALIGN",
          "30": "X86_FP80",
          "31": "PPC_FP128",
          "4": "DECLARE",
          "5": "DEFINE",
          "6": "PTR",
          "7": "GET_ELEMENT_PTR",
          "8": "INBOUNDS",
          "9": "X",
          "ALIGN": 3,
          "BFLOAT": 26,
          "CALL": 18,
          "CLOSE_BRACKET": 22,
          "CLOSE_CURLY_BRACKET": 24,
          "CLOSE_PARENTHESIS": 20,
          "COMMA": 15,
          "CONSTANT": 2,
          "DECLARE": 4,
          "DEFINE": 5,
          "DUOBLE": 28,
          "EQUALS": 11,
          "FLOAT": 27,
          "FP128": 29,
          "GET_ELEMENT_PTR": 7,
          "GLOBAL": 12,
          "HALF": 25,
          "INBOUNDS": 8,
          "NO_ALIAS": 14,
          "NO_CAPTURE": 13,
          "NO_FREE": 17,
          "OPEN_BRACKET": 21,
          "OPEN_CURLY_BRACKET": 23,
          "OPEN_PARENTHESIS": 19,
          "PPC_FP128": 31,
          "PRIVATE": 0,
          "PTR": 6,
          "RETURN": 10,
          "UNNAMED_ADDR": 1,
          "VOID": 16,
          "X": 9,
          "X86_FP80": 30,
        },
        "id": 1,
      },
      "type": Object {
        "enum": Object {
          "0": "IDENTIFIER",
          "1": "NUMBER_LITERAL",
          "2": "INTEGER_TYPE",
          "3": "STRING_LITERAL",
          "4": "KEY_WORD",
          "IDENTIFIER": 0,
          "INTEGER_TYPE": 2,
          "KEY_WORD": 4,
          "NUMBER_LITERAL": 1,
          "STRING_LITERAL": 3,
        },
        "id": 4,
      },
    },
    "type": 0,
  },
  Object {
    "data": undefined,
    "type": 1,
  },
  Object {
    "data": Object {
      "data": Object {
        "enum": Object {
          "0": "PRIVATE",
          "1": "UNNAMED_ADDR",
          "10": "RETURN",
          "11": "EQUALS",
          "12": "GLOBAL",
          "13": "NO_CAPTURE",
          "14": "NO_ALIAS",
          "15": "COMMA",
          "16": "VOID",
          "17": "NO_FREE",
          "18": "CALL",
          "19": "OPEN_PARENTHESIS",
          "2": "CONSTANT",
          "20": "CLOSE_PARENTHESIS",
          "21": "OPEN_BRACKET",
          "22": "CLOSE_BRACKET",
          "23": "OPEN_CURLY_BRACKET",
          "24": "CLOSE_CURLY_BRACKET",
          "25": "HALF",
          "26": "BFLOAT",
          "27": "FLOAT",
          "28": "DUOBLE",
          "29": "FP128",
          "3": "ALIGN",
          "30": "X86_FP80",
          "31": "PPC_FP128",
          "4": "DECLARE",
          "5": "DEFINE",
          "6": "PTR",
          "7": "GET_ELEMENT_PTR",
          "8": "INBOUNDS",
          "9": "X",
          "ALIGN": 3,
          "BFLOAT": 26,
          "CALL": 18,
          "CLOSE_BRACKET": 22,
          "CLOSE_CURLY_BRACKET": 24,
          "CLOSE_PARENTHESIS": 20,
          "COMMA": 15,
          "CONSTANT": 2,
          "DECLARE": 4,
          "DEFINE": 5,
          "DUOBLE": 28,
          "EQUALS": 11,
          "FLOAT": 27,
          "FP128": 29,
          "GET_ELEMENT_PTR": 7,
          "GLOBAL": 12,
          "HALF": 25,
          "INBOUNDS": 8,
          "NO_ALIAS": 14,
          "NO_CAPTURE": 13,
          "NO_FREE": 17,
          "OPEN_BRACKET": 21,
          "OPEN_CURLY_BRACKET": 23,
          "OPEN_PARENTHESIS": 19,
          "PPC_FP128": 31,
          "PRIVATE": 0,
          "PTR": 6,
          "RETURN": 10,
          "UNNAMED_ADDR": 1,
          "VOID": 16,
          "X": 9,
          "X86_FP80": 30,
        },
        "id": 2,
      },
      "type": Object {
        "enum": Object {
          "0": "IDENTIFIER",
          "1": "NUMBER_LITERAL",
          "2": "INTEGER_TYPE",
          "3": "STRING_LITERAL",
          "4": "KEY_WORD",
          "IDENTIFIER": 0,
          "INTEGER_TYPE": 2,
          "KEY_WORD": 4,
          "NUMBER_LITERAL": 1,
          "STRING_LITERAL": 3,
        },
        "id": 4,
      },
    },
    "type": 0,
  },
  Object {
    "data": undefined,
    "type": 1,
  },
  Object {
    "data": Object {
      "data": Object {
        "itemsType": Object {
          "data": 8,
          "type": Object {
            "enum": Object {
              "0": "FILE",
              "1": "DECLARE",
              "10": "IDENTIFIER",
              "11": "NUMBER",
              "12": "VOID_TYPE",
              "13": "BLOCK",
              "14": "RETURN_INSTRUCTION",
              "15": "CALL_ASSIGNABLE",
              "16": "ASSIGNMENT_INSTRUCTION",
              "17": "INPUT",
              "18": "INPUT_FLAG",
              "2": "FUNCTION",
              "3": "INTEGER_TYPE",
              "4": "GLOBAL_VARIABLE",
              "5": "STRING",
              "6": "FLOAT_TYPE",
              "7": "POINTER_TYPE",
              "8": "ARRAY_TYPE",
              "9": "STRUCT_TYPE",
              "ARRAY_TYPE": 8,
              "ASSIGNMENT_INSTRUCTION": 16,
              "BLOCK": 13,
              "CALL_ASSIGNABLE": 15,
              "DECLARE": 1,
              "FILE": 0,
              "FLOAT_TYPE": 6,
              "FUNCTION": 2,
              "GLOBAL_VARIABLE": 4,
              "IDENTIFIER": 10,
              "INPUT": 17,
              "INPUT_FLAG": 18,
              "INTEGER_TYPE": 3,
              "NUMBER": 11,
              "POINTER_TYPE": 7,
              "RETURN_INSTRUCTION": 14,
              "STRING": 5,
              "STRUCT_TYPE": 9,
              "VOID_TYPE": 12,
            },
            "id": 3,
          },
        },
        "length": 13,
      },
      "type": Object {
        "enum": Object {
          "0": "FILE",
          "1": "DECLARE",
          "10": "IDENTIFIER",
          "11": "NUMBER",
          "12": "VOID_TYPE",
          "13": "BLOCK",
          "14": "RETURN_INSTRUCTION",
          "15": "CALL_ASSIGNABLE",
          "16": "ASSIGNMENT_INSTRUCTION",
          "17": "INPUT",
          "18": "INPUT_FLAG",
          "2": "FUNCTION",
          "3": "INTEGER_TYPE",
          "4": "GLOBAL_VARIABLE",
          "5": "STRING",
          "6": "FLOAT_TYPE",
          "7": "POINTER_TYPE",
          "8": "ARRAY_TYPE",
          "9": "STRUCT_TYPE",
          "ARRAY_TYPE": 8,
          "ASSIGNMENT_INSTRUCTION": 16,
          "BLOCK": 13,
          "CALL_ASSIGNABLE": 15,
          "DECLARE": 1,
          "FILE": 0,
          "FLOAT_TYPE": 6,
          "FUNCTION": 2,
          "GLOBAL_VARIABLE": 4,
          "IDENTIFIER": 10,
          "INPUT": 17,
          "INPUT_FLAG": 18,
          "INTEGER_TYPE": 3,
          "NUMBER": 11,
          "POINTER_TYPE": 7,
          "RETURN_INSTRUCTION": 14,
          "STRING": 5,
          "STRUCT_TYPE": 9,
          "VOID_TYPE": 12,
        },
        "id": 8,
      },
    },
    "type": 2,
  },
  Object {
    "data": undefined,
    "type": 1,
  },
  Object {
    "data": Object {
      "data": "Hello World! ",
      "type": Object {
        "enum": Object {
          "0": "FILE",
          "1": "DECLARE",
          "10": "IDENTIFIER",
          "11": "NUMBER",
          "12": "VOID_TYPE",
          "13": "BLOCK",
          "14": "RETURN_INSTRUCTION",
          "15": "CALL_ASSIGNABLE",
          "16": "ASSIGNMENT_INSTRUCTION",
          "17": "INPUT",
          "18": "INPUT_FLAG",
          "2": "FUNCTION",
          "3": "INTEGER_TYPE",
          "4": "GLOBAL_VARIABLE",
          "5": "STRING",
          "6": "FLOAT_TYPE",
          "7": "POINTER_TYPE",
          "8": "ARRAY_TYPE",
          "9": "STRUCT_TYPE",
          "ARRAY_TYPE": 8,
          "ASSIGNMENT_INSTRUCTION": 16,
          "BLOCK": 13,
          "CALL_ASSIGNABLE": 15,
          "DECLARE": 1,
          "FILE": 0,
          "FLOAT_TYPE": 6,
          "FUNCTION": 2,
          "GLOBAL_VARIABLE": 4,
          "IDENTIFIER": 10,
          "INPUT": 17,
          "INPUT_FLAG": 18,
          "INTEGER_TYPE": 3,
          "NUMBER": 11,
          "POINTER_TYPE": 7,
          "RETURN_INSTRUCTION": 14,
          "STRING": 5,
          "STRUCT_TYPE": 9,
          "VOID_TYPE": 12,
        },
        "id": 5,
      },
    },
    "type": 2,
  },
]
`;
